# What is NumPy?

## Overview

[NumPy](https://numpy.org/) is the fundamental package for scientific computing in Python. It is a Python library that provides a multidimensional array object, various derived objects (such as masked arrays and matrices), and an assortment of routines for fast operations on arrays, including mathematical, logical, shape manipulation, sorting, selecting, I/O, discrete Fourier transforms, basic linear algebra, basic statistical operations, random simulation and much more.

## Learning Objectives

* Understand the difference between one-, two- and n-dimensional arrays in NumPy
* Understand how to apply some linear algebra operations to n-dimensional arrays without using for-loops
* Understand axis and shape properties for n-dimensional arrays

## Prerequisites

You’ll need to know a bit of Python. For a refresher, see the [Python tutorial](https://docs.python.org/tutorial/).

## Get Started

To get started with this module, you will need to ensure that numpy is installed.

In [98]:
import numpy as np
print(np.__version__)
# Install the 'numpy' package using pip (Python's package installer) if the above statement failed
# The '%pip' magic command is used in Jupyter notebooks to install packages directly from the notebook
# %pip install numpy

1.26.4


## Array Basics

NumPy’s main object is the homogeneous multidimensional array (ndarray). It is a table of elements (usually numbers), all of the same type, indexed by a tuple of non-negative integers. In NumPy dimensions are called *axes*.

### An example NumPy ndarray

In [99]:
# 'arange()' returns evenly spaced values within a given interval.
# Values are generated within the half-open interval [start, stop) 
# (in other words, the interval including start but excluding stop).
# Here, np.arange(15) generates an array with values from 0 to 14 (15 elements).
a = np.arange(15)

# 'reshape()' gives a new shape to an array without changing its data.
# In this case, we reshape the 1D array into a 2D array with 3 rows and 5 columns.
# The total number of elements must remain the same, i.e., 3*5 = 15 elements.
a = a.reshape(3, 5)

# Display the reshaped 2D array
a

array([[ 0,  1,  2,  3,  4],
       [ 5,  6,  7,  8,  9],
       [10, 11, 12, 13, 14]])

### Array attributes

#### `ndarray.ndim`

The number of axes of the array.

In [100]:
# Access the number of dimensions (axes) of the array `a`
# `ndim` is an attribute that returns the number of dimensions in a NumPy array.
# For example, if `a` is a 2D array, `a.ndim` would return 2.
a.ndim

2

#### `ndarray.shape`

This is a tuple of integers indicating the size of the array in each dimension. For a matrix with n rows and m columns, shape will be (n,m). The length of the shape tuple is therefore the number of axes, ndim.

In [101]:
# Get the shape of the array or dataframe 'a'
# This will return a tuple representing the dimensions (rows, columns) or (length,) depending on the type of 'a'
a.shape

(3, 5)

#### `ndarray.size`

The total number of elements of the array. This is equal to the product of the elements of shape.

In [102]:
# `a.size` returns the total number of elements in the array `a`
# It gives the product of the dimensions of the array
a.size

15

#### `ndarray.dtype`

An object describing the type of the elements in the array. One can create or specify dtypes using standard Python types. Additionally NumPy provides types of its own. numpy.int32, numpy.int16, and numpy.float64 are some examples.

In [103]:
# This retrieves the data type (dtype) of the variable or array 'a'
a.dtype

dtype('int64')

#### `ndarray.itemsize`

The size in bytes of each element of the array. For example, an array of elements of type float64 has itemsize 8 (=64/8), while one of type complex32 has itemsize 4 (=32/8). It is equivalent to ndarray.dtype.itemsize.

In [104]:
# a.itemsize returns the size (in bytes) of one element in the array 'a'
# This is useful for understanding how much memory each individual element of the array occupies.
a.itemsize

8

#### `type()`

The data type of the object.

In [105]:
# Get the type of the object 'a'
# The type() function returns the type of the specified object
type(a)

numpy.ndarray

### Creating Arrays

You can create an array from a regular Python list or tuple using the array function. The type of the resulting array is deduced from the type of the elements in the sequences.

In [106]:
# Importing the numpy library, which provides support for arrays and mathematical operations
import numpy as np

# Creating a numpy array with elements 2, 3, and 4
a = np.array([2, 3, 4])

# Display the numpy array 'a'
a

array([2, 3, 4])

In [107]:
# `a.dtype` returns the data type (dtype) of the NumPy array or pandas Series `a`
# It helps identify the type of data stored in the array (e.g., int, float, object, etc.)
a.dtype

dtype('int64')

In [108]:
# Adding .name extracts just the string name of the data type (e.g., 'int32', 'float64')
a.dtype.name

'int64'

In [109]:
# Create a NumPy array with three elements: 1.5, 3.5, and 5.5
b = np.array([1.5, 3.5, 5.5])

# Display the NumPy array
b

array([1.5, 3.5, 5.5])

In [110]:
# Get the data type of the object `b`
# `b.dtype` returns the type of the elements stored in the array-like object `b`
b.dtype

dtype('float64')

A frequent error consists in calling array with multiple arguments, rather than providing a single sequence as an argument.

In [111]:
a = np.array(1, 2, 3, 4)    # WRONG

In [112]:
a = np.array([1, 2, 3, 4])  # RIGHT

**array()** transforms sequences of sequences into two-dimensional arrays, sequences of sequences of sequences into three-dimensional arrays, and so on.

In [113]:
# Create a 2D NumPy array with 2 rows and 3 columns
b = np.array([[1, 2, 3], [4, 5, 6]])

# Display the created 2D array
b

array([[1, 2, 3],
       [4, 5, 6]])

The type of the array can also be explicitly specified at creation time

In [114]:
# Create a 2x2 numpy array with complex numbers
# The 'dtype=complex' ensures the array elements are of complex type
c = np.array([[1, 2], [3, 4]], dtype=complex)

# Display the array
c

array([[1.+0.j, 2.+0.j],
       [3.+0.j, 4.+0.j]])

The function **zeros** creates an array full of zeros, the function **ones** creates an array full of ones, and the function **empty** creates an array whose initial content is random and depends on the state of the memory. By default, the dtype of the created array is **float64**, but it can be specified via the key word argument **dtype**.

In [115]:
# Return a new array of given shape and type, filled with zeros.
# np.zeros() creates an array of the specified shape filled with zeros.
# The shape (4, 5) means the array will have 4 rows and 5 columns.

np.zeros((4, 5))

array([[0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0.]])

In [116]:
# Return a new array of the given shape and type, filled with ones.
# The shape of the array is (3, 4, 5), meaning it will have 3 blocks, 
# each containing a 4x5 matrix. 
# The dtype is specified as np.int16, which means each element in the array 
# will be of type 16-bit integer.
np.ones((3, 4, 5), dtype=np.int16)

array([[[1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1]],

       [[1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1]],

       [[1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1]]], dtype=int16)

In [117]:
# Return a new array of the given shape and type, filled with arbitrary data.
# The shape of the array is specified as (3, 4), meaning 3 rows and 4 columns.
np.empty((3, 4))

array([[0.14100771, 0.96298181, 0.62170053, 0.22127978],
       [0.91594584, 1.89594927, 0.10730838, 0.51304512],
       [0.37631448, 0.76193221, 0.55999372, 0.49371024]])

In [118]:
# Return the identity array of size 4x4.
# The identity matrix has ones on the diagonal and zeros elsewhere.
np.identity(4)

array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.]])

In [119]:
# Generate a 2x2 array filled with random numbers from a standard normal distribution
# (mean = 0, standard deviation = 1)
arr = np.random.randn(2, 2)

# Display the array
arr

array([[ 1.61346664, -1.40969133],
       [-1.82305731, -0.83827323]])

### Printing Arrays

One-dimensional arrays are then printed as rows, bi-dimensionals as matrices and tri-dimensionals as lists of matrices.

In [120]:
# Create a NumPy array 'a' with 24 elements, ranging from 0 to 23 (exclusive) using np.arange
a = np.arange(24)

# Print the created array 'a'
print(a)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20 21 22 23]


In [121]:
# Create a 1D numpy array with values from 0 to 23 (24 elements)
b = np.arange(24)

# Reshape the 1D array into a 2D array with 6 rows and 4 columns
b = b.reshape(6, 4)

# Print the resulting 2D array
print(b)

[[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]
 [12 13 14 15]
 [16 17 18 19]
 [20 21 22 23]]


In [122]:
# Create a NumPy array of shape (2, 3, 4) with values from 0 to 23
# np.arange(24) generates an array of 24 values from 0 to 23
c = np.arange(24).reshape(2, 3, 4)

# Print the reshaped array
print(c)

[[[ 0  1  2  3]
  [ 4  5  6  7]
  [ 8  9 10 11]]

 [[12 13 14 15]
  [16 17 18 19]
  [20 21 22 23]]]


### Basic Operations

Arithmetic operators on arrays apply elementwise. A new array is created and filled with the result.

In [123]:
# Create a NumPy array `a` with the specified values
a = np.array([10, 20, 30, 40])

# Display the array `a`
a

array([10, 20, 30, 40])

In [124]:
# Create a NumPy array with values ranging from 0 to 3
b = np.arange(4)

# Display the array
b

array([0, 1, 2, 3])

In [125]:
# Subtract the value of variable 'b' from variable 'a' and store the result in 'c'
c = a - b

# Output the value of 'c'
c

array([10, 19, 28, 37])

In [126]:
# Square the value of 'b' (raise 'b' to the power of 2)
b**2

array([0, 1, 4, 9])

In [127]:
# Checks if the value of 'a' is less than 40
a < 40

array([ True,  True,  True, False])

The product operator * operates elementwise in NumPy arrays. The matrix product can be performed using the @ operator or the dot function or method:

In [128]:
# Define a 2x2 NumPy array A
A = np.array([[1, 2],
              [3, 4]])

# Define a 2x2 NumPy array B
B = np.array([[5, 6],
              [7, 8]])

# Perform element-wise multiplication of matrices A and B
# Each element in A is multiplied by the corresponding element in B
A * B     # elementwise product
          # A = [[ a, b ],
          #      [ c, d ]]
          # B = [[ e, f ],
          #      [ g, h ]]
          # A * B = [[ ae, bf ],
          #          [ cg, dh ]]
          # a = 1, b = 2, c = 3, d = 4, e = 5, f = 6, g = 7, h = 8
          # ae = 1x5 = 5
          # bf = 2x6 = 12
          # cg = 3x7 = 21
          # dh = 4x8 = 32

array([[ 5, 12],
       [21, 32]])

In [129]:
A @ B     # Matrix multiplication using the "@" operator in Python
          # A = [[ a, b ],
          #      [ c, d ]]
          # B = [[ e, f ],
          #      [ g, h ]]
          # A @ B = [[ ae + bg, af + bh ],
          #          [ ce + dg, cf + dh ]]
          # a = 1, b = 2, c = 3, d = 4, e = 5, f = 6, g = 7, h = 8
          # ae + bg = 1x5 + 2x7 = 19
          # af + bh = 1x6 + 2x8 = 22
          # ce + dg = 3x5 + 4x7 = 43
          # cf + dh = 3x6 + 4x8 = 50

array([[19, 22],
       [43, 50]])

In [130]:
A.dot(B)  # Computes the matrix product of A and B using the dot product operation
          # A = [[ a, b ],
          #      [ c, d ]]
          # B = [[ e, f ],
          #      [ g, h ]]
          # A @ B = [[ ae + bg, af + bh ],
          #          [ ce + dg, cf + dh ]]
          # a = 1, b = 2, c = 3, d = 4, e = 5, f = 6, g = 7, h = 8
          # ae + bg = 1x5 + 2x7 = 19
          # af + bh = 1x6 + 2x8 = 22
          # ce + dg = 3x5 + 4x7 = 43
          # cf + dh = 3x6 + 4x8 = 50

array([[19, 22],
       [43, 50]])

+= and *=, act in place to modify an existing array rather than create a new one.

In NumPy, you can transpose an array using the .T attribute, the transpose() method, or the numpy.transpose() function.

In [131]:
A.T

array([[1, 3],
       [2, 4]])

In [132]:
# Create a 2x3 array filled with ones
# The dtype=int ensures that the elements are integers
a = np.ones((2, 3), dtype=int)

# Display the created array
a

array([[1, 1, 1],
       [1, 1, 1]])

In [133]:
# Create an instance of the default random number generator with a fixed seed (1) for reproducibility
rg = np.random.default_rng(1)  

# Generate a 2x3 array of random numbers uniformly distributed between 0 and 1
b = rg.random((2, 3))

# Display the generated array
b

array([[0.51182162, 0.9504637 , 0.14415961],
       [0.94864945, 0.31183145, 0.42332645]])

In [134]:
# Multiply the current value of 'a' by 3 and update 'a' with the new value
a *= 3  

# Display the updated value of 'a'
a

array([[3, 3, 3],
       [3, 3, 3]])

In [135]:
# Add the value of variable `a` to the value of variable `b`
b += a

# Return the updated value of `b`
b

array([[3.51182162, 3.9504637 , 3.14415961],
       [3.94864945, 3.31183145, 3.42332645]])

Many unary operations such as **sum**, **min**, **max** are also available.

In [136]:
# Create a 3x4 matrix of random values using a random number generator (rg)
# This generates a matrix with 3 rows and 4 columns, filled with random numbers
a = rg.random((3, 4))

# Display the generated matrix 'a'
a

array([[0.82770259, 0.40919914, 0.54959369, 0.02755911],
       [0.75351311, 0.53814331, 0.32973172, 0.7884287 ],
       [0.30319483, 0.45349789, 0.1340417 , 0.40311299]])

In [137]:
# Sum all the elements of the DataFrame or Series 'a'
a.sum()

5.517718775393902

In [138]:
# Get the minimum value from the array, pandas Series, or DataFrame 'a'
a.min()

0.027559113243068367

In [139]:
# Call the max() method on the variable 'a' to find the maximum value in it
# If 'a' is a NumPy array, pandas Series, or a similar data structure,
# this will return the largest value in 'a'.
a.max()

0.8277025938204418

By default, these operations apply to the array as though it were a list of numbers, regardless of its shape. However, by specifying the **axis** parameter you can apply an operation along the specified axis of an array:

In [140]:
# Create a NumPy array with values ranging from 0 to 23 (inclusive), with a total of 24 elements
# Then reshape the array into a 4x6 matrix (4 rows and 6 columns)
b = np.arange(24).reshape(4, 6)

# Display the reshaped matrix
b

array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11],
       [12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23]])

In [141]:
# Sum of each column in the DataFrame `b`
# The `axis=0` argument ensures that the sum is calculated for each column (vertically)
b.sum(axis=0)

array([36, 40, 44, 48, 52, 56])

In [142]:
# Calculate the minimum value of each row in the DataFrame `b`
b.min(axis=1)  # axis=1 indicates that the operation is performed row-wise

array([ 0,  6, 12, 18])

In [143]:
# Cumulative sum along each row (axis=1 means operation is performed row-wise)
b.cumsum(axis=1)

array([[  0,   1,   3,   6,  10,  15],
       [  6,  13,  21,  30,  40,  51],
       [ 12,  25,  39,  54,  70,  87],
       [ 18,  37,  57,  78, 100, 123]])

### Universal Functions (ufunc) in NumPy

In NumPy, a **Universal Function (ufunc)** is a function that operates element-wise on an array. **ufunc**s provide efficient vectorized operations and broadcasting capabilities, making computations faster and more concise compared to traditional Python loops.

#### Features of ufuncs

- **Element-wise operations:** Apply functions to each element of an array without explicit loops.
- **Vectorization:** Perform computations efficiently using optimized C code under the hood.
- **Broadcasting:** Enable operations on arrays of different shapes.
- **Type casting:** Automatically convert input types to appropriate data types.
- **Reduce and accumulate:** Perform cumulative operations on arrays.

In [144]:
# Create a NumPy array with values from 0 to 2 (3 is exclusive)
B = np.arange(3)

# Display the array B
B

array([0, 1, 2])

In [145]:
# Apply the exponential function to each element in the array or matrix B
# The np.exp() function calculates e raised to the power of each element in B
np.exp(B)

array([1.        , 2.71828183, 7.3890561 ])

#### How np.exp(B) Iterates Through Matrix Elements
The function np.exp(B) operates on the array element by element:

- **Step 1**: Takes B[0] = 0, computes exp(0) = 1.0
- **Step 2**: Takes B[1] = 1, computes exp(1) ≈ 2.71828183
- **Step 3**: Takes B[2] = 2, computes exp(2) ≈ 7.3890561

The entire computation is vectorized, meaning NumPy efficiently applies the operation without using explicit loops.

In [146]:
# Take the square root of the value stored in variable B using NumPy's sqrt function
np.sqrt(B)

array([0.        , 1.        , 1.41421356])

In [147]:
# Create a NumPy array `C` with values [2.0, -1.0, 4.0]
C = np.array([2., -1., 4.])

# Add the arrays `B` and `C` element-wise using NumPy's `add` function
# This will return a new array where each element is the sum of the corresponding elements of `B` and `C`
np.add(B, C)

array([2., 0., 6.])

### Indexing, Slicing and Iterating

One-dimensional arrays can be indexed, sliced and iterated over, much like lists and other Python sequences.

In [148]:
# Create a NumPy array of numbers from 0 to 9 (inclusive)
# The `np.arange(10)` generates an array with values from 0 to 9

a = np.arange(10)**3  # Raise each element of the array to the power of 3

# Display the resulting array
a

array([  0,   1,   8,  27,  64, 125, 216, 343, 512, 729])

In [149]:
a[2]  # Access the third element of the array 'a'

8

In [150]:
a[2:5] # get the third through fifth elements of the array (not including fifth)

array([ 8, 27, 64])

In [151]:
a[0:6:2] # from start to position 6, get every 2nd element (start:end:stepsize)

array([ 0,  8, 64])

In [152]:
a[::-1] # reversed array

array([729, 512, 343, 216, 125,  64,  27,   8,   1,   0])

In [153]:
for i in a: # use for loop to access every element of the array
  print(i**(1 / 3.))

0.0
1.0
2.0
3.0
3.9999999999999996
4.999999999999999
5.999999999999999
6.999999999999999
7.999999999999999
8.999999999999998


In [154]:
b

array([[ 0,  1,  2,  3,  4,  5],
       [ 6,  7,  8,  9, 10, 11],
       [12, 13, 14, 15, 16, 17],
       [18, 19, 20, 21, 22, 23]])

In [155]:
b[2, 3] # third row and fouth column element of b

15

In [156]:
b[0:5, 1] # each row in the second column of b

array([ 1,  7, 13, 19])

In [157]:
b[:, 1] # equivalent to the previous example

array([ 1,  7, 13, 19])

In [158]:
b[1:3, :] # the second and third row of b

array([[ 6,  7,  8,  9, 10, 11],
       [12, 13, 14, 15, 16, 17]])

Iterating over multidimensional arrays is done with respect to the first axis:



In [159]:
# Iterate through each element (row) in the iterable `b`
for row in b:
    # Print the current element (row) from the iterable `b`
    print(row)

[0 1 2 3 4 5]
[ 6  7  8  9 10 11]
[12 13 14 15 16 17]
[18 19 20 21 22 23]


However, if one wants to perform an operation on each element in the array, one can use the **flat** attribute which is an iterator over all the elements of the array:

In [160]:
# Iterate over each element in the flattened version of the array 'b'
# 'b.flat' returns an iterator over all the elements of 'b' in a flat (1-dimensional) order
for element in b.flat:
    # Print each individual element of the flattened array
    print(element)

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23


Arrays can be indexed by arrays of integers and arrays of booleans.

In [161]:
# Create a NumPy array containing the first 12 square numbers using np.arange() and squaring them
a = np.arange(12)**2  # np.arange(12) generates numbers from 0 to 11, and squaring them gives their square values

# Display the array
a

array([  0,   1,   4,   9,  16,  25,  36,  49,  64,  81, 100, 121])

In [162]:
# Create a NumPy array containing a list of indices
i = np.array([0, 1, 3, 8, 5])  # an array of indices

# Display the array of indices
i

array([0, 1, 3, 8, 5])

In [163]:
a[i]  # Accesses the element of list/array `a` at the index `i`

array([ 0,  1,  9, 64, 25])

### Singular Value Decomposition (SVD)

**Singular Value Decomposition (SVD)** is a fundamental matrix factorization technique in linear algebra. It decomposes a matrix \( A \) into three matrices:

$$
A = U \Sigma V^T
$$

Where:  
- $A$: The original matrix (of size $m \times n$).  
- $U$: An orthogonal matrix (of size $m \times m$) containing the left singular vectors.  
- $\Sigma$: A diagonal matrix (of size $m \times n$) containing the singular values (non-negative and in descending order).  
- $V^T$: The transpose of an orthogonal matrix (of size $n \times n$) containing the right singular vectors.

#### Key Applications of SVD:
1. **Dimensionality Reduction**: Used in Principal Component Analysis (PCA) to reduce the number of features while preserving variance.
2. **Data Compression**: Reduces the size of data by approximating the original matrix with fewer singular values.
3. **Noise Reduction**: Removes noise by truncating smaller singular values.
4. **Recommendation Systems**: Used in collaborative filtering (e.g., Netflix recommendations).
5. **Image Processing**: Compresses and reconstructs images.

#### Example in Python:


In [164]:
import numpy as np  # Import the NumPy library

# Create a 3x3 matrix A
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

# Perform Singular Value Decomposition (SVD)
# U: Left singular vectors
# S: Singular values
# VT: Transpose of the right singular vectors (V^T)
U, S, VT = np.linalg.svd(A)

# Print the left singular vectors (U matrix)
print("U (Left Singular Vectors):\n", U)

# Print the singular values (S matrix, will be a 1D array)
print("S (Singular Values):\n", S)

# Print the transpose of the right singular vectors (VT matrix)
print("VT (Right Singular Vectors Transpose):\n", VT)


U (Left Singular Vectors):
 [[-0.21483724  0.88723069  0.40824829]
 [-0.52058739  0.24964395 -0.81649658]
 [-0.82633754 -0.38794278  0.40824829]]
S (Singular Values):
 [1.68481034e+01 1.06836951e+00 1.47280825e-16]
VT (Right Singular Vectors Transpose):
 [[-0.47967118 -0.57236779 -0.66506441]
 [-0.77669099 -0.07568647  0.62531805]
 [ 0.40824829 -0.81649658  0.40824829]]


### Array creation, filtering, and modification

In [165]:
# Create a NumPy array containing the names of three animal species
animals = np.array(["Tiger", "Elephant", "Cheetah"])

# Generate a 3x4 array of random numbers sampled from a standard normal distribution (mean=0, std=1)
# This array represents measurements associated with the animals (e.g., weight, length, speed, age)
animal_data = np.random.randn(3, 4)

# Display the generated animal data
print("Original animal data:")
print(animal_data)

Original animal data:
[[-0.49212024 -1.20740601  0.27009821 -1.12887281]
 [-0.06897505  0.04502148 -0.23409612  1.42436451]
 [-1.20380148 -1.04690335  0.37139877 -1.40528863]]


In [166]:
# Filter the `animal_data` array to retrieve the row(s) corresponding to the animal "Elephant"
# `animals == "Elephant"` creates a boolean array where the condition is True for "Elephant" and False otherwise
# This boolean array is used to index `animal_data`, returning the row(s) where the condition is True
print("\nData for Elephant:")
print(animal_data[animals == "Elephant"])


Data for Elephant:
[[-0.06897505  0.04502148 -0.23409612  1.42436451]]


In [167]:
# Filter the `animal_data` array to retrieve all elements that are less than -0.5
# `animal_data < -0.5` creates a boolean array of the same shape as `animal_data`, where each element is:
# - True if the corresponding element in `animal_data` is less than -0.5
# - False otherwise
# The boolean array is used to index `animal_data`, returning a flattened array of elements that satisfy the condition
print("\nElements less than -0.5:")
print(animal_data[animal_data < -0.5])


Elements less than -0.5:
[-1.20740601 -1.12887281 -1.20380148 -1.04690335 -1.40528863]


In [168]:
# Replace all elements in `animal_data` that are greater than 1 with 1
# `animal_data > 1` creates a boolean array of the same shape as `animal_data`, where each element is:
# - True if the corresponding element in `animal_data` is greater than 1
# - False otherwise
# The boolean array is used to index `animal_data`, and all elements satisfying the condition are set to 1
animal_data[animal_data > 1] = 1

# Display the modified `animal_data` array
print("\nModified animal data (values > 1 set to 1):")
print(animal_data)


Modified animal data (values > 1 set to 1):
[[-0.49212024 -1.20740601  0.27009821 -1.12887281]
 [-0.06897505  0.04502148 -0.23409612  1.        ]
 [-1.20380148 -1.04690335  0.37139877 -1.40528863]]


## Conclusion

In this module, we've learned about the basics of NumPy, including creating and manipulating arrays, using universal functions, and indexing/slicing arrays. These skills form the foundation for scientific computing in Python.

## Clean up

Remember to shut down your Jupyter Notebook instance when you're done to avoid unnecessary charges. You can do this by stopping the notebook instance from the Amazon SageMaker console.